### ISSUE TITLE
RFC: default error handling of Epics

### ISSUE THREAD

**Issue by jayphelps** [redux-observable/redux-observable#94]
The middleware is intentionally designed to accept a single, root Epic. This means that all your Epics are typically composed together using `combineEpics(...epics)`.

The problem with this approach is that if an exception is thrown inside one of your epics and it isn't caught by you, it will "bubble" up all the way to the root epic and cause the entire output stream to terminate aka `error()`. As an example, `Observable.ajax()` currently throws exceptions for non-200 responses. So one bad AJAX request can terminate all of your epics and your app no longer listens for actions to handle side effects.

You can demo this sort of behavior here: http://jsbin.com/yowiduh/edit?js,output notice that it works the first time, errors the second, and any subsiquent time does nothing because the epic has terminated!

This is not unlike most programming models (where an uncaught exception terminates the "process"). In fact, this behavior is the same as if you were doing the equivalent in imperative JS with generators or similar. So RxJS is _"correct"_ in the implementation.

However, because this has critical implications, we've been discussing adding some default behavior to mitigate this; especially for people who may be fairly new to Rx and not realize the implications.

Here are a couple options:
1. `.catch()` it and and resubscribe to the root epic, basically "restarting" epics to continue listening.
   - Epics that maintain internal state will have lost that state completely, **even ones that didn't cause the exception.**
2. or have `combineEpics()` listen for individual epics throwing exceptions and restart only the one that produced the error.

Alternative suggestions are welcome.

---

If we proof-of-concept option 1. it might look something like this:

http://jsbin.com/kalite/edit?js,output

``` js
const output$ = rootEpic(action$, store).catch((error, source) => {
  Observable.throw(error, Scheduler.async).subscribe();
  return source;
});

output$.subscribe(store.dispatch);
```

I'm not sure we want to do `Observable.throw(error, Scheduler.async).subscribe();` but I used it here because I do want the exception to be thrown so `window.onerror` sees it and it shows up in the console as expected, but we also need to return the original source observable so I'm using `Scheduler.async` to "defer" the exception. There's prolly a better way to do this. A simple setTimeout prolly is ok instead.

Cc/ @blesh

**Comment by jayphelps** [comment]
I've decided on do option 1 for now, feedback welcome. PR to come shortly.

**Comment by jayphelps** [comment]
Turns out I was wrong saying that option 1 would allow unrelated epics to maintain their internal state. Not sure why I thought that in hindsight. Effectively option 1 _is in fact_ still letting the entire root epic terminate and then "restarting" it entirely with a new subscription.

I've updated [the jsbin](http://jsbin.com/kalite/edit?js,output) to reflect this--the console will show an incrementing number which will reset back to zero whenever the other ajax epic errors.

Again, that means option 1, catching errors at the root epic level, would cause **all* epics to lose their internal state, and restart.

Option 2 is catching them individually inside `combineEpics`:

```js
export const combineEpics = (...epics) => (...args) =>
  merge(
    ...epics.map(epic =>
      epic(...args)
        // When epic emits error, rethrow it async and resubscribe.
        // Otherwise, errors will cause cascading termination of all epics. 
        ::_catch((error, source) => {
          setTimeout(() => { throw error; }, 0);
          return source;
        })
    )
  );
```

This however has its own caveats.

* If you don't use `combineEpics`, obviously it won't catch them--we could _also_ catch the root epic as well, so that seems like a doable solution?
* Having `combineEpics` do this feels like magic that will most often go unknown to people and confuse them why this behavior doesn't happen when they invoke epics manually when doing higher-order epics aka composition.

```js
const epic1 = (action$, store) => stuff();
const epic2 = (action$, store) => epic1(moreStuff(), store);
```

We could later mitigate this by introducing a paved way to "fork" epics, as we've been experimenting with for a while. Forking is just like composition except you don't need to pass along the store cause it'll be passed along for you. We could catch errors in forks as well.

* Lastly, this would break some people who are right now catching their own errors and logging them. We've actually helped several people do this, and it would be particularly onerous because if they wanted to continue they'd have to effectively re-create `combineEpic` that _doesn't have this catching behavior.

```js
const rootEpic = (action$, store) =>
  combineEpics(epic1, epic2)(action$, store)
    .catch((e, source) => {
      logTheErrorToAPIOrSomething(e);
      return source;
    });
```

**Comment by slorber** [comment]
Related problems are experienced by redux-saga users regularly.

I think the solution is to implement "supervision strategy" that actor systems have. Sagas are generally implemented on the backend by using actor systems like Akka (even Greg Young, best advocate of eventsourcing/DDD said Akka is a saga framework).

The idea of the actor systems resilience is that the system should be able to heal itself on its own with supervision strategy. If one part of the actor tree fails, we can declare which strategy to adopt for that part of the tree.

The epic restarting itself is actually one of the most used supervision strategy in actor systems.

I think the behavior should rather not be encoded directly inside combineEpics but rather as a wrapper for each epic, because one should be able to setup fine-grained behavior per epic

More infos on this issue: https://github.com/yelouafi/redux-saga/issues/570

**Comment by jayphelps** [comment]
@slorber people can opt into automatic retry already today with the [`.retry()` operator](http://reactivex.io/rxjs/class/es6/Observable.js~Observable.html#instance-method-retry):

```js
const somethingEpic = action$
  action$.ofType(SOMETHING)
    .mergeMap(action => Observable.throw('BAD'))
    .retry();
```

Perhaps the answer to this whole problem is just to make this operator + usage more obvious in the docs...but I dunno.

This whole effort was mostly because I felt people would practically _always_ want an epic to restart if it errors out. The argument against I guess is that those errors are uncaught exceptions for a reason--we can't guarantee the behavior of the app continues correctly if we always restart them. However, in JS world I would argue people are very much used to apps which may have errors but try to chug along anyway, even if some things are not broken. In the redux-observable case, if a single epic errors out without being caught, _all_ of their epics will terminate. This cascade is pretty dubious.

**Comment by slorber** [comment]
This is the exact same problem with redux-saga: people have a root saga and fork child sagas, but an error in a single child saga does kill the root saga, and thus the other child sagas, (so there's no saga left in the end). This happened to me because of bad connections: a single http request failure could make my whole app unusable. There's however a way to know it crashed as we can receive uncaught errors in a callback and display something to the user.

Still think it should be fine-grained, as sometimes you want the error to propagate to the parent (which may retry or propagate also), and sometimes you want it to retry. Generally, at the very top of the tree, you want to retry as the sagas/epics are generally unrelated and work in isolation, but inside an isolated saga, you may want to propagate to the parent so that it's the parent that does the retry.


About RxJS retry(), according to the marble, will it resubscribe to past events as well? Because a Saga should conceptually not receive twice the same events. In backend systems, sagas/process managers generally use at-least-once semantics to be sure to receive the async messages, but make sure to handle it only once in any case by keeping reference/version/whatever of already handled messages. Not sure to understand if retry() is really suitable to the problem because the marble seems to resubscribe to the whole stream. Not doing RxJS for a long time so maybe it's just a problem of type of source observable (subject/hot/cold) ?

**Comment by jayphelps** [comment]
@slorber the marble is displaying the behavior with the assumption that the source is lazy/cold. So resubscribing causes it to start producing as-new. Most Observables are this way, but Subjects are hot/multicast.

The ActionsObservable we provide epics is a little confusing to think about deeply in that itself is lazy/cold but it subscribes internally to a Subject which is hot. The main take away is that for any practical reason, resubscribing to `action$` will _not_ replay any previous actions. So using `.retry()` will subscribe to any _future_ actions from that point on. In the case of redux-observable epics using it on the top-level, the term "retry" is a bit of misnomer and instead can be thought of as "ignore the error and restart the epic"

**Comment by slorber** [comment]
I see. I guess you keep a hot action stream for event replay / hot code reloading, and make epic stream cold on purpose to avoid receiving past events on failure. Makes sense to me.

So yes I think this pattern should rather be documented instead of a default retry strategy in combineEpics which would probably be too magical and not flexible in case you don't want this behavior

**Comment by jayphelps** [comment]
@slorber 

> I guess you keep a hot action stream for event replay / hot code reloading, and make epic stream cold on purpose to avoid receiving past events on failure. Makes sense to me.

Not sure I follow? To clarify: if you use `.retry()` on the top-level stream returned in an epic, it will _not_ replay previous actions. On error, it will simply "restart" the epic, listening for _future_ actions.

The real stream of actions that redux-observable maintains internally is hot, which is why previous actions are not replayed--by design.

**Comment by blesh** [comment]
Are we looking for a way to sandbox individual epics? I think that happens anyhow.

### If we restart or otherwise "handle" unhandled errors by default

**Benefits**
1. Probably the most ergonomic for the most users.
2. Developers don't have to learn proper error handling with observables.

**Rishs**
1. Sometimes users will want their epic to die in the event of an error, or errors. Think of repeated ajax errors or something, perhaps they want it to just die after 100 attempts.
2. Blanket catching errors that are unknown to developers might squelch feedback they're otherwise expecting.
3. It's harder to "opt-out" of automagic error handling
4. Perhaps harder to control strategy. immediate retry? incremental step back?
5. Developers don't have to learn proper error handling with observables.

### If we don't handle any errors

**Benefits**
1. Developers can configure their error handling in whatever way they choose.
2. Developers forced to learn error handling techniques in Rx.

**Risks**
1. Might be less ergonomic for the general use case where most developers do not want there epics to die after an error.
2. Developers forced to learn error handling techniques in Rx.

**Comment by jayphelps** [comment]
just jotting down some rough "create epic helper" ideas to provide some default behavior (including restarting on error), with escape hatches via optional config arg.

```js
// having options first seems the most visually appealing,
// but is an awkward overload if they don't provide options
const somethingEpic = createEpic({ restartOnError: false }, action$ =>
  action$.ofType(PING)
    .map(() => pong())
);

// so if default behavior is fine, first arg is epic implementation
// itself. but again..overloads are awkward
const somethingEpic = createEpic(action$ =>
  action$.ofType(PING)
    .map(() => pong())
);

// putting options last means we don't need an
// overload, but it sure looks awkward for arrow fns
// without a block
const somethingEpic = createEpic(action$ =>
  action$.ofType(PING)
    .map(() => pong()),
  { restartOnError: false } 
);
```

**Comment by rgbkrk** [comment]
I'm definitely in favor of redux-observable not handling any errors for the reasons Ben mentioned:

> 1. Developers can configure their error handling in whatever way they choose.
> 2. Developers forced to learn error handling techniques in Rx.

I'd rather see recipes for error handling, exponential backoff, and retrying.

**Comment by slorber** [comment]
same for me. It's also the path taken by @yelouafi on redux-saga btw

**Comment by jayphelps** [comment]
If we go that route, I'm struggling to decide how the docs should prescribe when to use `.retry()` and when not to? For many, it seems like they'd want to add it to the end of every one of their epics..which seems pretty annoying. Thoughts?

**Comment by rgbkrk** [comment]
@harph sung high praise for the `examples/navigation` section. I think more in-context examples would be good.

I feel bad that all the examples I can provide are so far in the weeds of a domain area that they would not be helpful to users (unless they're hacking on our code).

**Comment by BerkeleyTrue** [comment]
@jayphelps I'm also in favor of this

**Comment by blesh** [comment]
I think maybe RxJS can provide an interesting operator or two around this problem to help. Requires more thought though.

**Comment by salanki** [comment]
As a data point, this is how we do error handling in our (react native) app. A custom `combineEpics` that adds a `repeat()` to all epics.
```
return output$.repeat();
```

The point of the `repeat()` is to allow error catching in the epic with `catch(x => Observable.of(errorActionCreator(x)))`. This pattern can then be used on both epics (hot observables) and cold ones that would just end up in an endless loop if adopting the `catch((x, stream) => Observable.merge(Observable.of(errorActionCreator(x)), stream))` pattern everywhere.

On our top level combined epic we add a catch-all just as described earlier in this thread. This catch-all is only to be used as a last resort. All epics are expected to take care of their own errors.

**Comment by jayphelps** [comment]
@salanki I believe using `repeat()` in that way would prevent your epics from shutting down gracefully in a server-render, but most people don't do SSR anyway. Just something to note.

**Comment by meticoeus** [comment]
I'm also in favor of redux-observable not handling errors, but I think it would be really helpful if it would log uncaught exceptions so you can find out which epic is letting errors slip through.

I've also found that it silently crashes when exceptions are thrown in reducers, making simple syntax errors hard to track down.

**Comment by jayphelps** [comment]
@meticoeus right now no errors should be swallowed by redux-observable. Errors in your reducers should also not cause any errors or other interactions with your epics. If this is happening and you're confident it's redux-observable, could you create an issue with a way to reproduce?

You can see that exceptions bubble all the way here: http://jsbin.com/konebug/edit?js,output

**Comment by jtomaszewski** [comment]
Thank you guys for all the comments. I've got annoyed by mi epics breaking too often in my dev environment, so using your tips, here's some code that others might use:

```js
// lib/helpers/epic.js
export const wrapEpic = epic => (...args) =>
  epic(...args).catch((error, source) => {
    setTimeout(() => { throw error }, 0)
    return source
  })

// modules/auth/epic.js
import { combineEpics } from "redux-observable"
import { wrapEpic } from "lib/helpers/epic"

export const epic = combineEpics( ...[
  epic1,
  epic2,
].map(wrapEpic) )

// app/rootEpic.js
import { combineEpics } from "redux-observable"
import { wrapEpic } from "lib/helpers/epic"

import authEpic from "modules/auth/epic"

export default combineEpics( ...[
  authEpic,
].map(wrapEpic) )
```

I added the `wrapEpic` to every place I use `combineEpics` and now all the epics handle the errors properly.

IMHO if I'd want my epic to stop and never restart, then instead of throwing an error, there are also other ways do it (f.e. https://github.com/Reactive-Extensions/RxJS/blob/master/doc/api/core/operators/never.md ) .

**Comment by mmcgahan** [comment]
@jayphelps I think I'm having the same issue that @meticoeus is having with redux-observable appearing to swallow errors and crash the root epic.

I've slightly tweaked the JSBin you shared to put the error in the reducer

http://jsbin.com/rosiyapali/edit?js,console

The error gets thrown and hits `window` so it's definitely not being swallowed, but I also don't know how to catch it - it doesn't call `onError` in the `pingEpic` so I can't retry/restart. I feel like I must be missing something obvious, but it might be a case where redux-observable doesn't provide the right hook to handle the error

**Comment by jayphelps** [comment]
@mmcgahan oh I see, redux-observable definitely is not intended to manage your reducers and any exceptions they may throw. 😄 You can do that yourself with some third party middleware or making your own.

http://jsbin.com/xefopiw/edit?js,console

Just make sure to put it _before_ the redux-observable middleware (or any other one really) because otherwise the errors won't be caught.

**Comment by mmcgahan** [comment]
ah that makes a lot of sense, thanks!

**Comment by gorshkov-leonid** [comment]
@jtomaszewski 
After catch error on root epics, combine latest will not work.... :
https://github.com/Reactive-Extensions/RxJS/issues/1036

**Comment by jayphelps** [comment]
@gorshkov-leonid not sure what you mean--btw that link refers to RxJS v4, v5 is here https://github.com/ReactiveX/rxjs and was a complete rewrite

**Comment by gorshkov-leonid** [comment]
@jayphelps I not sure too :thinking:  
Example with usage of 5.1.1: https://jsfiddle.net/goleon/hq2c15cf/
I emulated 'save' with global error handler and 'load' with error handler inside switch map. May be problem in switch map...?  So [advice|https://redux-observable.js.org/docs/recipes/ErrorHandling.html] to handle error inside mergeMap seems logical...

As result root handler does not guarantee working condition... Why he is needed then? То show blue critical modal dialog with message 'please refresh page'?

**Comment by jayphelps** [comment]
@gorshkov-leonid ah yes, your example shows what I needed. Indeed you need to isolate your Observable chains, otherwise you allow that error to propagate out which terminates the `action$.ofType` and any parent epic (like your root epic) until it either reaches a catch or bubbles all the way up and no epic is listening at all.

**Comment by gorshkov-leonid** [comment]
Interesting analogy. I saw the code of retry operation. After error handler will will be made new subscription. Is it possible to extend this idea? Developer can set uncatchedErrorsHandler, that provides ability to decide how to handle errors. But subscription in any case will be restored.... Is it fantastic?

**Comment by jayphelps** [comment]
@gorshkov-leonid not sure I follow. Can you provide pseudo code for what you mean?

**Comment by braco** [comment]
@jayphelps 

> Indeed you need to isolate your Observable chains, otherwise you allow that error to propagate out which terminates the action$.ofType and any parent epic (like your root epic) until it either reaches a catch or bubbles all the way up and no epic is listening at all.

Ahhhhh, fuck. Noob here. This is a pretty brutal open door and should probably be foregrounded more heavily in the documentation and have a noisier output in development mode, don't you think? I was combining epics from several files and tracked the problem down to an error in a unrelated (so I thought) file. At the very least, maybe just "unhandled exception" in console.warn?

**Comment by jayphelps** [comment]
@braco it's not clear how we could improve developer experience or documentation around this without trying to teach them RxJS, which our docs currently try to warn people away from using redux-observable if they don't know Rx. Any suggestions?

> At the very least, maybe just "unhandled exception" in console.warn?

Hmm. This should already happen, unless you swallow the error somewhere. redux-observable nor Rx will swallow errors silently, unless you do so yourself.

**Comment by braco** [comment]
> redux-observable nor Rx will swallow errors silently, unless you do so yourself.

Wasn't intentionally swallowing any errors, will check on that. Thank you. If an error can be suppressed and still propagate, I wonder if something like [redux-saga's warning](https://github.com/redux-saga/redux-saga/blob/f9dfb587ed2ec3f564d143ac448197bc89f2c137/src/internal/proc.js#L315) would be a possible addition.

This thread touches on hand holding vs Rx convention, but since this seems more about redux/sugar, why not present some optional safety?

> // Plural since each epic is its own root
> rootEpics([epics], config)
>
> isolate: true // Children can't propagate errors
> restartOnError: true, // [Like this](https://github.com/redux-observable/redux-observable/issues/94#issuecomment-263730326)
> throwOnError: true // [Like this](https://github.com/redux-observable/redux-observable/issues/94#issuecomment-277351312)

@jayphelps, thanks for the patience you seem to express toward everyone, it's noticed and appreciated.

**Comment by jayphelps** [comment]
> If an error can be suppressed and still propagate, I wonder if something like redux-saga's warning would be a possible addition.

I'm not familiar with that particular part of redux-saga. Can you maybe describe how it would work in redux-observable land? If it's similar to your comment below, you can ignore this question.

> This thread touches on hand holding vs Rx convention, but since this seems more about redux/sugar, why not present some optional safety?

Some patterns have been discussed above, including the possibility of providing some by default, but the latest consensus seemed to be that most preferred redux-observable let users decide--but I think documenting these various patterns in the Recipes section would be super super helpful, I just haven't had time. 😄 

That said, I don't have strong opinions on this as long as it remains flexible still (e.g. opt out of default error handling). Maybe you can convince the others here who have opposed? hehe

> thanks for the patience you seem to express toward everyone, it's noticed and appreciated.

You're welcome!💃 That's for being so understanding.

**Comment by gorshkov-leonid** [comment]
I do not understand the implementation. But I can agree with one point. Yes, the main criterion to which one should strive is safety. 
Ah yes I noticed this issue in "redux-observable", I think solution can be build on top of rxjs feature. How to do it? If chain unsubscribe on error somewhere inside rxjs, what can we do?

**Comment by braco** [comment]
@jayphelps 

> Can you maybe describe how [redux-saga's warning] would work in redux-observable land?

In dev mode, if an epic failed or was cancelled from propagation, there would be a console.info note. Every epic would announce its own cancellation. As the error propagated, you'd get `epic {name} was cancelled` for each epic, and probably `root epic was cancelled`. It was vital with redux-saga to catch problems. Would also be great if there was a mini-stack if possible: `epic Bar was cancelled (origin: Foo)`.

If unadorned RxJS is important, noisier dev would be a reasonable compromise imo, and would go a long way to guide users in self-diagnosis. Flow control is nasty business, and I would assume this library will usually be the gateway drug to RxJS instead of the opposite.

I do think some default safety is warranted given the specificity in Redux convention. Even if that was just a default `warnUncaughtErrors: true` and optional `throwUncaughtErrors: true` that wrapped root and reducer-level epics like @jtomaszewski's code above.

You could also sneak in some RTFM hints in console output like React is wont to do: "epic Foo was cancelled. Make sure to catch errors, as .... see: http://reactivex.io/rxjs/...", but not sure how generalized that can be since you're library agnostic?

Which of you oppositional interlocutors needs convincing?

**Comment by jayphelps** [comment]
is @braco and @gorshkov-leonid the same person or different? It's hard to tell 😄

**Comment by braco** [comment]
😄 We're 2 cautionary ghosts, as in A Christmas Carol. Ghost of noob future is coming. Ignore us at your own risk

**Comment by jayphelps** [comment]
> In dev mode, if an epic failed or was cancelled from propagation, there would be a console.info note. Every epic would announce its own cancellation. As the error propagated, you'd get epic {name} was cancelled or epic {name} ended for each epic, and probably root epic was cancelled. It was vital with redux-saga to catch problems. Would also be great if there was a mini-stack if possible: epic Bar was cancelled (origin: Foo).

That seems great, generally. It is a good compromise to use `console.error` to log about the epics terminating to help debug, but still bubble and cascade the error to let people handle it how they choose.

Anyone have any objections?

**Comment by BerkeleyTrue** [comment]
@jayphelps Where would this happen, `combineEpics` or `createEpicMiddleware`?

**Comment by jayphelps** [comment]
@BerkeleyTrue depends. I experimented a bit and if someone is using `combineEpics` we can use it to annotate the error message with the originating epic's name but still bubble it up and then console.error it in the middleware if it reaches that, with something like `"all epics terminated due to uncaught error in myAwesomeEpic"` but still let the error bubble further, naturally resulting in `window.onerror` triggering and the true error shown in the console.

**Comment by jayphelps** [comment]
I'm hesitant to also do console.error inside `combineEpics` itself because

* if it bubbled to the root, there would be three total errors in the console
  1. original
  2. X epic terminated (from combineEpics)
  3. all epics terminated because of X epic. (from middleware)
* just because an epic terminated from the perspective of combineEpics doesn't mean they won't catch the error before it reaches the root epic, so the error message would be annoying to some.

**Comment by BerkeleyTrue** [comment]
I like the idea of annotating the error message in combineEpics. 

```js

import { merge } from 'rxjs/observable/merge';

import { errorSymbol } from './error-symbol.js';

function annotate(err, epic) {
  const epicName = epic.name || 'Anon Epic';
  // prevent error source from overriding
  if (err[errorSymbol]) {
    // could even be an additive process
    // like so: err[errorSymbol] = `${epicName}(${err[errorSymbol]})`;
    return err;
  }
  err[errorSymbol] = epicName;
  return err;
};
/**
  Merges all epics into a single one.
 */
export const combineEpics = (...epics) => (...args) =>
  merge(
    ...epics.map(epic => {
      const output$ = epic(...args);
      if (!output$) {
        throw new TypeError(`combineEpics: one of the provided Epics "${epic.name || '<anonymous>'}" does not return a stream. Double check you\'re not missing a return statement!`);
      }
      return output$.catch(err => [annotate(err, epic.name)]);
    })
  );
```

**Comment by BerkeleyTrue** [comment]
@jayphelps I've also been thinking about a different approach that could help out during development.

What do you think of the addition of a new function, say `createEpic`, That will take an epic and add a bunch of additional checks to the epic and in production default to an identity function?

**Comment by jayphelps** [comment]
@BerkeleyTrue I've played with similar things but nothing ended up being super useful _to me_ but I'm biased cause I know how everything works. :clown_face: having more meaningful checks/errors would be helpful to newbies I imagine so it's worth considering. If you have time to put together a proof of concept, feel free to open an RFC or even PR if it's easy.

I will say I think we'd still allow regular functions-as-epics as it is today. For me, one of the major benefits of the existing approach is that really it's just a plain old function so you can do normal functional things for custom abstractions, etc. and it's simpler to learn IMO. But I'll try to keep an open mind with any proposals.

The original goal was minimal amount of glue between redux + RxJS, with the expectation that the community might naturally create additional, more opinionated abstractions that people could choose to adopt or not.

**Comment by meticoeus** [comment]
I think anything redux-observable does for you to catch errors or report on errors should be opt-in and well documented. I prefer the flexibility of RxJS to handle any types of errors appropriate to the situation that might throw them.

@braco 

> If unadorned RxJS is important, noisier dev would be a reasonable compromise imo, and would go a long way to guide users in self-diagnosis. Flow control is nasty business, and I would assume this library will usually be the gateway drug to RxJS instead of the opposite.

As long as this library is a minimalistic glue layer between redux and RxJS (which I hope it remains) I think users are going to find it much easier to learn RxJS without the added complexity of interacting with redux and then learning to use them together, rather than using redux-observable as the gateway direction.

**Comment by jayphelps** [comment]
@meticoeus Just to be clear, would you be opposed to the default (not opt-in) behavior of: errors that caused the termination of _every_ epic were `console.error` logged, but not caught? So a friendly message would let you know all epics were terminated, but the original error itself would still get caught by things like `window.onerror` etc.

I think this would be useful in every situation, even for Rx pros.

**Comment by meticoeus** [comment]
@jayphelps That sounds reasonable.

**Comment by braco** [comment]
If this helps any rank amateurs in the meantime, I'm doing this:

https://gist.github.com/braco/28dc2006e9ce04a004d2d9c4f5b9a4b6

Convention here is

epics.js
```
import reducerName1 from 'reducerName1/epics';
...

export default wrapEpics({ reducerName1 });
```

reducerName1/epics.js
```
export const epicName = epic$ => ...
...
```

Then you get debug messages like `reducerName/epicName errored with ...` and `reducerName/epicName exited with ...`

**Comment by lsentkiewicz** [comment]
This is my wrapper for unit testing (Jest)

```js
const wrapEpic = epic => (...args) =>
  epic(...args).catch((error) => {
    console.error(error);
    expect.assertions(Infinity);
  });
```
It logs an error with correct stack trace, and makes sure the test always fails.  
Adding extra `throw error` doesn't make any difference.

**Comment by nhducit** [comment]
@jayphelps I think we should mention this issue in Readme and provide a workaround solution.

**Comment by hally9k** [comment]
@jayphelps I am trying to use the adapter option in `createEpicMiddleware` to add a catchError to every epic in my epic tree. I'm getting a bit confused about how the pieces fit together, a little help? Is this approach possible. It seems (theoretically) cleaner that the above approach...

**Comment by jayphelps** [comment]
@hally9k I would recommend instead using function composition in your root Epic:

Assuming rxjs v6 and redux-observable 1.0.0-beta.1:

```js
// TODO: replace doSomethingAndReturnObservable
const combineEpicsAndCatchErrors = (...epics) =>
  return (action$, state$) => {
    epics = epics.map(epic => (action$, state$) => epic(action$, state$).pipe(
      catchError(e => doSomethingAndReturnObservable(e))
    ));
    return combineEpics(...epics)(action$, state$);
  };
};

const rootEpic = combineEpicsAndCatchErrors(...epics);
```

If you don't need to do it individually for each epic, it's easier to just do it to the rootEpic (but has different behavior):

```ts
const rootEpic = (action$, state$) =>
  combineEpics(...epics)(action$, state$).pipe(
    catchError(e => doSomethingAndReturnObservable(e))
  );
````

**Comment by GodefroyClair** [comment]
This is a quite long discussion to swallow... Any place where we can find a sumup of the recommended practice?

**Comment by marceloemanoel** [comment]
Hey guys, any update on this?

**Comment by TheMoonDawg** [comment]
We've actually just had this behavior burn us on one of our production applications. We have a lot of users in remote areas with shotty internet connectivity where this is happening more often than not. We _think_ failed requests that are returning undefined to the stream are to blame, but it's hard to tell when the app just silently fails.

@jayphelps Would you still recommend the last solution you gave on June 12th, 2018 as the best way to approach rescuing our application whenever it fails?

Thanks for all of the hard work you've done on this amazing library!

**Comment by tylerweitzman** [comment]
Also suffering this issue here. Some bugs are also difficult to trace because they fail silently. Then the whole of the epics stop functioning.

We have too many epics to wrap individually.

I am trying to implement a solution into our HMR set up. The error is getting caught but (1) I cannot locate where the error happened (2) Epic isn't restarting on failure. I imagine I can call the same functionality that the HMR uses in order to reload the epic, even re-requiring the epic and then calling next, but I need to first kill the previous epic using store.dispatch, which doesn't yet exist when I create the root epic.

Here's my code
```
const composeEnhancers = composeWithDevTools({ serialize: true, trace: traceFlag, traceLimit: 25 });

const epic$ = new BehaviorSubject(rootEpic);

const hotReloadingEpic = (action$, ...rest) =>
    epic$.pipe(
        mergeMap(epic =>
            epic(action$, ...rest).pipe(
                catchError((error, caught) => {
                    const failedAction = Observable.of({
                        type: 'FAILED_EPIC',
                        location: 'ROOT',
                        caught,
                        error,
                    });
                    return failedAction
                }),
                takeUntil(action$.pipe(
                    ofType('HOT_RELOAD_EPICS')
                ))
            ))
    );

const epicMiddleware = createEpicMiddleware(hotReloadingEpic);

export const store = createStore(
    reducer,
    composeEnhancers(applyMiddleware(epicMiddleware, ...((process.env.ENV === 'DEV' || process.env.FORCE_DEV) && !process.env.DISABLE_REDUX_LOG) ? [logger] : [])),
);

if (module.hot) {
    module.hot.accept('./app.epic', () => {
        // eslint-disable-next-line
        const nextRootEpic = require('./app.epic').default;
        // First kill any running epics
        store.dispatch({ type: 'HOT_RELOAD_EPICS' });
        // Now setup the new one
        epic$.next(nextRootEpic);
    });
}
```

**Comment by jayphelps** [comment]
@tylerweitzman Hello!

> (1) I cannot locate where the error happened
I'm afraid I can't be a huge help here without admittedly a ton of specifics to your code and your debugger. But some things to keep in mind in case you hadn't: some browsers have better debuggers than others, and JavaScript has multiple ways errors can propagate and be caught at. e.g. uncaught rejected Promises do not trigger `window.onerror` instead `window.onunhandledrejection` but even that isn't implemented by all browsers. Not all browsers implement `window.onerror` consistently either, etc, etc. Personally, I think Chrome provides the best overall debugging experience currently for async.

Modern browsers will carry along a `error.stack` property containing the call stack, which may or may not get printed if you log the error. Some browsers used to overwrite that stack however if you rethrow it, not sure if that's still the case.

> (2) Epic isn't restarting on failure.
Are you saying it isn't happening when using the code you provided below, regarding HMR? Just wanna double check.

If so, you'll need to "restart" the epic yourself when you catch the error--in general RxJS terms you need to resubscribe to the source. Error handling with Observables is very much like like try/catch with functions. Even if you catch an error it won't continue on where it was because the error is exceptional, so the app is now in a state that potentially isn't safe to continue from. A developer can choose to restart/resubscribe, just like they can choose to call the function that threw an exception again, but that's a choice the programmer has to make realizing that doing so may be risky in the sense that your code is no longer in the state it was because it was bad. Am I making any sense? haha sorry, hard to explain sometimes.

Besides manually calling `source.subscribe()`, _which we want to avoid as much as possible_, a couple idiomatic ways of resubscribing in RxJS are using `retry()`, `repeat()`, or returning the source inside a  `catchError()`. In this context, using the last one is probably what you're looking for. If you want to 1. catch an error from the source 2. emit one last thing 3. then resubscribe to the source (starting over again)

Stripped down version:

```js
// Using this to simulate some sort of source that will throw
// an error after 1 second. It could be any Observable,
// including the one your root epic returns.
const fakeEpic = () => timer(1000).pipe(
  tap(() => {
    throw new Error('some error');
  })
);

const result$ = fakeEpic().pipe(
  catchError((error, source$) => {
    // source$ is the same Observable `fakeEpic()` returned
    // it just gives us an easy way to get a reference to it
    return merge(
      of('its OK I caught it: ' + error.message),
      source$
    );
  })
);

result$.subscribe({
  next: value => console.log(value),
  error: e => console.error('bad stuff:', e)
});
```

The key part is:

```js
merge(
  of('its OK I caught it: ' + error.message),
  source$
)
```
We're merging an Observable of one thing, our customized error string, with the original source. Because Observables are lazy, they're just like functions in that you can "call" or subscribe to them as many times as you want and they'll start from the beginning again. We could also have used `concat()` instead of `merge()` because the first Observable `of()` is synchronous, same thing in this specific case.

It's really key to remember though: **This will resubscribe to the source, not pick up where we left off**. So your app can and likely is in a bad state, if your app contains non-trivial Rx code. Will this matter if you're only using it for local code development HMR? Maybe, maybe not.

I will say I don't personally recommend people use HMR on Epics (or anything with side effects) because I've worked on apps in the past and wasted hours debugging issues that were only caused by me doing HMR on epics. It won't always be obvious when its the cause.

**Comment by connor4312** [comment]
For us, a way to handle errors, for telemetry if nothing else, is quite valuable. Silently breaking the application is not too great. We've put together a small library which implements a supervisor pattern as some have discussed above, and also supports restarting epics. This is useful for more complex SPA, where if one component of the application fails, you might want to try giving it a kick instead of allowing it to remain broken: https://github.com/mixer/epic-supervisor

**Comment by tylerweitzman** [comment]
@jayphelps Thanks so much for the detailed explanation. That was super helpful for me in gaining a better understanding.

@connor4312 Nice framework! Works well

My remaining issue (@connor4312 this is with your framework), is that the source map only seems to translate the very last stack trace, instead of the full trace. We're using `inline-source-map` with webpack devtool.

Here's a screenshot
![image](https://user-images.githubusercontent.com/2078974/56448063-275b0f00-62c1-11e9-9b53-6591608b1d9a.png)

**Comment by connor4312** [comment]
From my experience stacktraces in rxjs are generally pretty terrible. epic-supervisor doesn't do anything with the stacktraces it sees on errors, it just passes them along to you, and there's little we can do to fix them up.

We've been using epic-supervisor for a few days, which has so far been able to root out a handful of errors/edge cases on its flight towards production. Generally in our application we've kept epics pretty small, so many instances of errors have been easy to diagnose with the error message and epic source included in the context object (epic names being destroyed by minification in production build). But certainly not all cases will be simple, and I'm interested in ways to improve that telemetry. 

Something I've considered doing is including the last-sent action in the error context. This wouldn't always be relevant for asynchronous errors since actions may have been dispatch while the async task was being performed, but it may be useful in some scenarios. Definitely open to feedback and ideas (and pull requests, of course :D)

**Comment by tylerweitzman** [comment]
So the sourcemap not translating properly is an rxjs issue? not a redux-observable issue and not a epic-supervisor issue? At which level is this stack trace being generated / manipulated? (I don't think it's just the standard JS engine because when I just throw an error in a random imperative function the stack trace is perfect with all file names mapped from bundle).

**Comment by connor4312** [comment]
Not having sourcemaps in console.log'd stacktraces appears to be a Chrome issue https://bugs.chromium.org/p/chromium/issues/detail?id=376409. Sorry for not linking that the first time, I was on mobile.

**Comment by fnpro** [comment]
@jayphelps Okay, it seems that you mentioned "unhandled ajax call errors," but it seems that redux-observable can "crash" for errors having nothing to do with redux-observable/RxJS. Please consider the example:

```javascript
(action$, state$) =>
    action$.pipe(
      ofType('FIRST_ACTION'),
      map(action => ({
        type: 'NEW_ACTION',
        place: action.undefinedDeepProp.undefinedDeepProp // <- Error! All of redux-observable is down.
      }))
    ),
```

Then, of course, we get the error message, "Cannot read property 'undefinedDeepProp' of undefined." **To me, redux-observable crashing is a big deal and the risk goes beyond unhandled ajax call errors.** In addition, these types of issues could be conditional:

```javascript
(action$, state$) =>
    action$.pipe(
      ofType('FIRST_ACTION'),
      map(action => ({
        type: 'NEW_ACTION',
        place: action.isSomething ? action.otherProp : action.undefinedDeepProp.undefinedDeepProp // <- Error! All of redux-observable is down.
      }))
    ),
```

**What I suggest is that you at least put Option 2 in the Error Handling page under recipes in the docs, and you can tell people that they can use this new combineEpics in place of the official combineEpics:**

```javascript
const combineEpics = (...epics) => (...args) =>
  merge(
    ...epics.map(epic =>
      epic(...args).pipe(
        catchError((error, source) => {
          setTimeout(() => {
            throw error;
          }, 0);
          return source;
        })
      )
    )
  );
```

I know not everyone might use combineEpics, but probably a lot of people do, so I would appreciate it if at least us combineEpics people were helped out.

So, I am new to redux-observable/RxJS, but I have written about 800 lines of redux-observable/RxJS. When I was getting into it, I watched [Ben Lesh's talk "RxJS 5 Thinking Reactively,"](https://www.youtube.com/watch?v=3LKMwkuK0ZE&t=1217) which was very helpful in understanding how observables die. So, I previously tried writing a global error handler, but was unable to get it to work. Then today, I found this GitHub issue, which I am very thankful for. **The point: if I had not found this GitHub issue, I would have probably had to remove redux-observable from my project and go back to thunks or something else, because I build a somewhat mission-critical application for our company, and I cannot have the app crash just because one epic was written incorrectly.**

**Comment by jayphelps** [comment]
@fnpro Hey there! Sorry for the trouble! I did want to give you some feedback that the wording of your message came off as fairly aggressive and isn't productive with me. Admittedly, those sorts of comments make me want to help _less_ rather than more, which I'm sure was not the intended effect. ✌️

**Comment by jayphelps** [comment]
All--it's still unclear what the community at large would like done here, but I'm open if folks rally around one of them.

Perhaps the most minimum thing we should do is:

Catch all epic errors that reach the middleware (aren't caught by your epics) and log a warning in dev mode about how uncaught errors "terminate" your entire epic chain and that you might be interested in the various ways you can handle that--linking out to a new docs page that includes a few, but not all, examples of ways you might want to deal with it. We'd then rethrow the error we caught, so it still shows up in the console/logging infra.

If you have an alternative suggestion, please comment and folks can 👍 or 👎 as a way to vote and get insight into who's still following this discussion and cares.

**Comment by jayphelps** [comment]
Perhaps the most minimum thing we should do is:

Catch all epic errors that reach the middleware (aren't caught by your epics) and log a warning in dev mode about how uncaught errors "terminate" your entire epic chain and that you might be interested in the various ways you can handle that--linking out to a new docs page that includes a few, but not all, examples of ways you might want to deal with it. We'd then rethrow the error we caught, so it still shows up in the console/logging infra.

If you have an alternative suggestion, please comment and folks can 👍 or 👎 as a way to vote and get insight into who's still following this discussion and cares.

**Comment by meticoeus** [comment]
I think logging a warning that points you to the docs would be the least obtrusive approach, so that users can solve the problem using standard Rx tools. Providing guidance on best practices for this in the docs (on the existing error handling page) would be hugely helpful, such as capturing and logging errors, then restarting an epic in the event of unexpected failure (syntax errors, unhandled NPEs, etc).

**Comment by fnpro** [comment]
@jayphelps I am completely sorry if you thought my words were aggressive. Please believe me that I was not trying to be aggressive at all! I use **bold** and such to show importance, not anger. So, I updated my above comment so that it is more "down to earth." Thank you for all that you do!

**Comment by aplassen** [comment]
**Perhaps an eslint plugin** would be of help? Linting would ensure that you have proper error handling on every single epic. An analogy from Java is where you are forced to catch exceptions when you use a certain method (e.g. `public void writeSomeStuff() throws IOException`), ensuring that exceptions are handled. 

**In React it is kind of the same,** the uncaught error would be quite devastating for the application if left unhandled by the developer. To mitigate this, react provide so-called [error bounderies](https://reactjs.org/docs/error-boundaries.html) that lets you, mutch like try-catch, handle this where it makes sense.

**In redux-observable**, if you wanted to map uncaught exception and -- for example --  dispatch them to a general application-level Redux action, you could wrap them with something like:

```js
const fetchUsers = applicationEpic(($action) => (
     // ... 
)}) // Dispatches { type: 'APP_UNCAUGHT_ERROR' } when 
`

**Comment by sajdre** [comment]
I used the following approach:

1) Added epics$ as a third ("dependencies") parameter to make it possible to load epics dynamically

```
const getEpicsSubject = () => new BehaviorSubject(combineEpics(...epicsToCombine))

const epic$ = getEpicsSubject()

const getRootEpic = () => (action$, state$, dependencies) => epic$.pipe(
    mergeMap(epic => epic(action$, state$, dependencies))
)

const epicsMiddleware = createEpicMiddleware({
    dependencies: epic$
})
```

2) Wrapped every epic before combine to make it "selfreloadable", so we can reload the same epic and just send message about error:

```
 const wrapEpicsToReloadOnFail = R.map(epic => {
    const selfReloadableEpic = (action$, state$, epics$) =>
        epic(action$, state$, epics$).pipe(
            catchError(error => {
                epics$.next(selfReloadableEpic)
                return of(ACTION_ERROR("" + error))
            })
        )
    return selfReloadableEpic
})

const epicsToCombine = wrapEpicsToReloadOnFail([...yourEpics])
```

R.map - it's just ramdajs map function

3) Also I send ACTION_UNDO when ACTION_ERROR happens, so "undoable" parts of my store can return to the latest safe state. What is "the latest safe state" and how I manage undo/redo actions is another story, and the whole ACTION_UNDO step is optional, of course. Let me know if you need details. So my reducers creation looks like this: 

```
const reducers = combineReducers({
    stateA: makeUndoRedoable(stateA),
    stateB,
    stateC,
    stateD: makeUndoRedoable(stateD),
    etc...
})
```

If you thing it's a bad solution I will appreciate if you let me know why.
